data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/prepper/peerforwarder/PeerForwarder.java

```diff
@@ -85,69 +93,70 @@ public PeerForwarder(final PluginSetting pluginSetting) {
+        final Map<String, List<Span>> spansByTraceId = PeerForwarderUtils.splitByTrace(spans);
         // Group ResourceSpans by consistent hashing of traceId
-        for (final Record<ExportTraceServiceRequest> record : records) {
-            for (final ResourceSpans rs : record.getData().getResourceSpansList()) {
-                final List<Map.Entry<String, ResourceSpans>> rsBatch = PeerForwarderUtils.splitByTrace(rs);
-                for (final Map.Entry<String, ResourceSpans> entry : rsBatch) {
-                    final String traceId = entry.getKey();
-                    final ResourceSpans newRS = entry.getValue();
-                    final String dataPrepperIp = hashRing.getServerIp(traceId).orElse(StaticPeerListProvider.LOCAL_ENDPOINT);
-                    groupedRS.computeIfAbsent(dataPrepperIp, x -> new ArrayList<>()).add(newRS);
-                }
-            }
+        final Map<String, List<Span>> spansByEndPoint = new HashMap<>();
+        for (final Map.Entry<String, List<Span>> entry: spansByTraceId.entrySet()) {
+            final String traceId = entry.getKey();
+            final String dataPrepperIp = hashRing.getServerIp(traceId).orElse(StaticPeerListProvider.LOCAL_ENDPOINT);
+            spansByEndPoint.computeIfAbsent(dataPrepperIp, x -> new ArrayList<>()).addAll(entry.getValue());
         }
 
-        final List<Record<ExportTraceServiceRequest>> recordsToProcessLocally = new ArrayList<>();
-        final List<CompletableFuture<Record>> forwardedRequestFutures = new ArrayList<>();
+        final List<Record<Span>> recordsToProcessLocally = new ArrayList<>();
+        final Map<CompletableFuture<ExportTraceServiceRequest>, List<Span>> forwardedRequestFuturesToSpans = new HashMap<>();
 
-        for (final Map.Entry<String, List<ResourceSpans>> entry : groupedRS.entrySet()) {
+        for (final Map.Entry<String, List<Span>> entry : spansByEndPoint.entrySet()) {
             final TraceServiceGrpc.TraceServiceBlockingStub client = getClient(entry.getKey());
+            if (isLocalClient(client)) {
+                recordsToProcessLocally.addAll(entry.getValue().stream().map(Record::new).collect(Collectors.toList()));
+                continue;
+            }
 
             // Create ExportTraceRequest for storing single batch of spans
             ExportTraceServiceRequest.Builder currRequestBuilder = ExportTraceServiceRequest.newBuilder();
-            int currSpansCount = 0;
-            for (final ResourceSpans rs : entry.getValue()) {
-                final int rsSize = PeerForwarderUtils.getResourceSpansSize(rs);
-                if (currSpansCount >= maxNumSpansPerRequest) {
+            List<Span> currBatchSpans = new ArrayList<>();
+            for (final Span span : entry.getValue()) {
+                final ResourceSpans rs;
+                try {
+                    rs = oTelProtoEncoder.convertToResourceSpans(span);
```

Comment:
rs can be deplared in the try statement. It no longer appears to be used outside of the try statement.

```
try {
    final ResourceSpans rs = oTelProdoEncoder.convertToResourceSpans(span);
```